Remove an expensive call to flatten in `ndarray.init*`

mseri opened the pull request:
I hope this will mitigate https://github.com/owlbarn/owl/issues/453 (@pveber would you mind checking, I cannot run the benchmark at the moment).

Please be careful when reviewing. What was the reason of going through flatten in the first instance?

mseri commented on src/owl/dense/owl_dense_ndarray_generic.ml:
```diff
@@ -145,26 +145,24 @@ let flatten x = reshape x [|numel x|]
+  let n = Array.fold_right (fun c a -> c * a) d 1 in
+  let y = Array1.create k c_layout n in
   let s = Owl_utils.calc_stride d in
   let j = Array.copy s in
```
@ryanrhymes why do you need this copy?

ryanrhymes replied:
if user function `f` modifies `s`, then the algorithm breaks; by passing a copy, we are safe no matter what `f` does to `j`.

mseri replied:
Ah Yes, good point!

mseri commented:
I have found a moment to run the benchmarks.

With this patch:

```
❯ dune exec ./owl_perfs.exe mat-mat-mul-small
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌──────────────────────┬──────────┬─────────┬────────────┐
│ Name                 │ Time/Run │ mWd/Run │ Percentage │
├──────────────────────┼──────────┼─────────┼────────────┤
│ lacaml-mat-mat-mul-4 │ 352.58ns │  18.00w │     51.70% │
│ owl-mat-mat-mul-4    │ 681.94ns │ 148.00w │    100.00% │
└──────────────────────┴──────────┴─────────┴────────────┘


❯ dune exec ./owl_perfs.exe vec-init-small
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌───────────────────┬──────────┬─────────┬────────────┐
│ Name              │ Time/Run │ mWd/Run │ Percentage │
├───────────────────┼──────────┼─────────┼────────────┤
│ lacaml-vec-init-4 │ 133.54ns │   9.00w │     43.99% │
│ owl-vec-init-4    │ 303.57ns │  28.00w │    100.00% │
└───────────────────┴──────────┴─────────┴────────────┘
```

Without this patch:
```
❯ dune exec ./owl_perfs.exe vec-init-small
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌───────────────────┬──────────┬─────────┬────────────┐
│ Name              │ Time/Run │ mWd/Run │ Percentage │
├───────────────────┼──────────┼─────────┼────────────┤
│ lacaml-vec-init-4 │ 133.63ns │   9.00w │     38.13% │
│ owl-vec-init-4    │ 350.44ns │  32.00w │    100.00% │
└───────────────────┴──────────┴─────────┴────────────┘

~/code/tests/owl-perfs master 22s
❯ dune exec ./owl_perfs.exe vec-init-large
Estimated testing time 20s (2 benchmarks x 10s). Change using '-quota'.
┌─────────────────────┬────────────┬─────────┬────────────┐
│ Name                │   Time/Run │ mWd/Run │ Percentage │
├─────────────────────┼────────────┼─────────┼────────────┤
│ lacaml-vec-init-100 │   399.91ns │   9.00w │     31.22% │
│ owl-vec-init-100    │ 1_280.81ns │  32.00w │    100.00% │
└─────────────────────┴────────────┴─────────┴────────────┘
```

mseri commented:
On https://discuss.ocaml.org/t/working-with-a-huge-data-chunks/3955/10 I have read
```
Creating a view, e.g., via reshape or sub is usually slow, because it is a custom value and has to be created in the major heap, thus every creating of a bigarray triggers the minor collection. We were addressing those issues by preparing all the views that we need in advance (or having a pool of views). [...]
```
Maybe this is really the culprit and we need to find an alternative mechanism.

pveber commented:
Thanks for this attempt @mseri! I was wondering: the underlying type for matrices could be a sum type, that is a `Mat.t` could be:
```ocaml
type t =
  | Gen of (...) Genarray.t
  | Mat2 of (...) Array2.t
  | ...
```
When using only lapack operations, we could then avoid all overhead while keeping the API unchanged (and simple).

On a completely orthogonal note, I have found *extremely* useful to [type vectors and matrices operations in a rather strict way](https://github.com/biocaml/phylogenetics/blob/b7ca01042bb53ba09158307063f8c0237faca9f6/lib/linear_algebra.mli), it showed me lots of bugs compared to using `owl` API directly.

mseri commented:
> On a completely orthogonal note, I have found _extremely_ useful to [type vectors and matrices operations in a rather strict way](https://github.com/biocaml/phylogenetics/blob/b7ca01042bb53ba09158307063f8c0237faca9f6/lib/linear_algebra.mli), it showed me lots of bugs compared to using `owl` API directly.

I think it is worth opening an issue about this (See now https://github.com/owlbarn/owl/issues/480). I have also been coming back and forth to this issueFrom time to time.

mseri commented:
> the underlying type for matrices could be a sum type, that is a `Mat.t` could be:
> 
> ```ocaml
> type t =
>   | Gen of (...) Genarray.t
>   | Mat2 of (...) Array2.t
>   | ...
> ```
> 
> When using only lapack operations, we could then avoid all overhead while keeping the API unchanged (and simple).

This is an interesting idea. It may be worth to try and give it a go at a certain stage. We are in principle free to rebind operations in the modules with more efficient ones. Lots of this code could actually be auto-generated with a suitable ppx

nilsbecker commented:
An orthogonal dimension where typing could in principle help improve numerical efficiency is to include matrix structure in the types. LAPACK/BLAS have specialized routines for diagonalizing/decomposing e.g. symmetric or hermitian matrices. some matrix functions are known to preserve these properties, which could also be propagated in their (GAD)type in principle. e.g. eigenvalues of hermitian matrices are real. i think there is/was a ruby numerical library that did something like this, but could not confirm this now when googling Numo.

of course, a somewhat overlapping issue is that matrices with a certain structure can also be stored in various optimal ways (e.g. bands, only one half of symmetric matrices, etc). afaik it's an open question whether a system that tries to automatically dispatch matrices to/from different sparse storage formats according to their structure, can be useful. it might be doing more harm than good by making memory usage unpredictable.
